Sources/diag.swift

```diff
@@ -3,24 +3,102 @@
- *  Unless required by applicable law or agreed to in writing, software distributed under the 
- *  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either 
- *  express or implied. See the License for the specific language governing permissions and 
+ *  Unless required by applicable law or agreed to in writing, software distributed under the
+ *  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing permissions and
  *  limitations under the License.
  *
- *  Copyright 2016 Philip Erickson
+ *  Copyright 2018 Tor Rafsol Løseth
  **************************************************************************************************/
 
- public func diag() -> ()
- {
-     // FIXME: implement this
-     fatalError("Not yet implemented")
- }
\ No newline at end of file
+
+/// Returns a square diagonal matrix with the elements of vector v on the main diagonal,
+/// and zeros everywhere else.
+///
+/// - Parameter v   : A vector data structure
+/// - Returns: A matrix data structure
+public func diag(v: Vector<Double>) -> Matrix<Double>
+{
+    return diag(v: v, k: 0)
+}
+
+
+/// Returns a square diagonal matrix with the elements of vector v on the kth diagonal,
+/// and zeros everywhere else.
+///
+/// - Parameters:
+///   - v: A vector data structure.
+///   - k: The kth diagonal, k = 0 is the main diagonal, k > 0 is above the main diagonal, and k < 0 is below the main diagonal.
+/// - Returns: A square matrix data structure.
+public func diag(v: Vector<Double>, k: Int) -> Matrix<Double>
+{
+    
+    let length = v.count + abs(k)
+    let A = Matrix(length, length, value: 0.0)
+    var data = A.data
+    
+    // Diagonals have indices intervals at every other:
+    let interval = A.rows + 1
+    
+    let diagonalStartIndex = k >= 0 ? k : length * abs(k)
+    
+    let vectorIndices = (0..<v.count).map {index in
+        diagonalStartIndex + (index * interval)
+    }
+    
+    vectorIndices.enumerated().forEach {(index, value) in
+        data[value] = v.data[index]
+    }
+    
+    return Matrix(length, length, data)
+}
+
+
+/// Returns a vector data structure with the elements from the main diagonal of a square matrix.
+///
+/// - Parameter A: A square matrix data structure.
+/// - Returns: A vector structure with diagonal elements.
+public func diag(A: Matrix<Double>) -> Vector<Double>
+{
+    return diag(A: A, k: 0)
+}
+
+
+/// Returns a vector data structure with the elements from the kth diagonal of a square matrix.
+///
+/// - Parameters:
+///   - A: A square matrix data structure.
+///   - k: The kth diagonal, k = 0 is the main diagonal, k > 0 is above the main diagonal, and k < 0 is below the main diagonal.
+/// - Returns: A vector data structure.
+public func diag(A: Matrix<Double>, k: Int) -> Vector<Double>
+{    
+    precondition(A.rows == A.columns, "Matrix must be square")
```

Comment:
I think we eventually want to move away from using a raw `precondition` or `assert` to use the wrappers `warn` and `error`. I was hoping that the wrappers would help in testing but I kind of got stuck. We don't have to change this now, but perhaps it's a good time to have a conversation about this; I'd love to pick your brain